src/OrchardCore.Modules/OrchardCore.Workflows/Controllers/WorkflowTypeController.cs

```diff
@@ -127,17 +118,16 @@ public async Task<IActionResult> Index(WorkflowTypeIndexOptions options, PagerPa
                 .Take(pager.PageSize)
                 .ListAsync();
 
-            // The existing session's connection is returned, don't dispose it
-            var connection = await _session.CreateConnectionAsync();
-
             var dialect = _session.Store.Configuration.SqlDialect;
             var sqlBuilder = dialect.CreateBuilder(_session.Store.Configuration.TablePrefix);
             sqlBuilder.Select();
             sqlBuilder.Distinct();
             sqlBuilder.Selector(nameof(WorkflowIndex), nameof(WorkflowIndex.WorkflowTypeId), _session.Store.Configuration.Schema);
             sqlBuilder.Table(nameof(WorkflowIndex), alias: null, _session.Store.Configuration.Schema);
 
-            var workflowTypeIdsWithInstances = await connection.QueryAsync<string>(sqlBuilder.ToSqlString());
+            // Use existing session connection. Do not use 'using' or dispose the connection.
+            var connection = await _session.CreateConnectionAsync();
+            var workflowTypeIdsWithInstances = (await connection.QueryAsync<string>(sqlBuilder.ToSqlString(), param: null, _session.CurrentTransaction)).ToList();
```

Comment:
These changes look OK but how do they fix the issue?

Reply:
I don't know if it did or not. The issue in production happens randomly which is really hard to reproduce. But this is the only suspicious code since it's the only place to try to reuse the current connection. In production, the issue seems to happen when we list the workflow types. I think the fact we could be writing data to the current transaction and then read on the same connection seems that we should be passing the transaction.

Reply:
Hmm, OK, this won't hurt, but I really don't see what here would fix the exception. BTW this might be related: https://github.com/OrchardCMS/OrchardCore/issues/15290.